### PR TITLE
KIP-1068: New metrics for the new KafkaConsumer

### PR DESCRIPTION
**Pull request by brenden20**
This PR will implement KIP-1068, adding new metrics to the consumer for the CONSUMER rebalance protocol.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 1]
```diff
@@ -320,10 +322,11 @@ private void process(final ConsumerRebalanceListenerCallbackNeededEvent event) {
 
             ApiVersions apiVersions = new ApiVersions();
             final BlockingQueue<ApplicationEvent> applicationEventQueue = new LinkedBlockingQueue<>();
-            final BackgroundEventHandler backgroundEventHandler = new BackgroundEventHandler(backgroundEventQueue);
+            final BackgroundEventHandler backgroundEventHandler = new BackgroundEventHandler(backgroundEventQueue, kafkaConsumerMetrics);
 
             // This FetchBuffer is shared between the application and network threads.
             this.fetchBuffer = new FetchBuffer(logContext);
+            final NetworkThreadMetricsManager networkThreadMetricsManager = new NetworkThreadMetricsManager(metrics);
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:329]: construct it in the ApplicationEventHandler

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java [hunk 2]
```diff
@@ -67,14 +69,18 @@ public class ConsumerNetworkThread extends KafkaThread implements Closeable {
     private final IdempotentCloser closer = new IdempotentCloser();
     private volatile Duration closeTimeout = Duration.ofMillis(DEFAULT_CLOSE_TIMEOUT_MS);
     private volatile long cachedMaximumTimeToWait = MAX_POLL_TIMEOUT_MS;
+    // Visible for testing
+    final NetworkThreadMetricsManager metricsManager;
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java]: i couldn't find any usage of this metrics manager.  If this needs to be accessed in test maybe write a getter instead.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java [hunk 3]
```diff
@@ -84,6 +90,8 @@ public ConsumerNetworkThread(LogContext logContext,
         this.networkClientDelegateSupplier = networkClientDelegateSupplier;
         this.requestManagersSupplier = requestManagersSupplier;
         this.running = true;
+        this.metricsManager = metricsManager;
+        this.usingMetrics = (metricsManager != null);
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java]: i'm not sure why we need a variable like this.  Make the metricsManager optional if it is not always there.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventHandler.java [hunk 4]
```diff
@@ -54,13 +57,37 @@ public ApplicationEventHandler(final LogContext logContext,
...
                 applicationEventReaper,
                 applicationEventProcessorSupplier,
                 networkClientDelegateSupplier,
-                requestManagersSupplier);
+                requestManagersSupplier,
+                null);
+        this.networkThread.start();
+    }
+
+    public ApplicationEventHandler(final LogContext logContext,
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventHandler.java:74]: just modify the original constructor and the factor class should work

Comment (philipnee) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventHandler.java:74]: This is Method Reference. so you don't need to add an extra ctor

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java (hunk 3 above)
Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java]: Yes, `Optional<NetworkThreadMetricsManager>` is neater.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/BackgroundEventHandler.java [hunk 5]
```diff
@@ -30,9 +32,24 @@
...
+    private final KafkaConsumerMetrics kafkaConsumerMetrics;
+    private final KafkaShareConsumerMetrics kafkaShareConsumerMetrics;
+    private boolean isShareConsumer;
 
-    public BackgroundEventHandler(final Queue<BackgroundEvent> backgroundEventQueue) {
+    public BackgroundEventHandler(final Queue<BackgroundEvent> backgroundEventQueue, KafkaConsumerMetrics kafkaConsumerMetrics) {
         this.backgroundEventQueue = backgroundEventQueue;
+        this.kafkaConsumerMetrics = kafkaConsumerMetrics;
+        this.kafkaShareConsumerMetrics = null;
+        isShareConsumer = false;
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/BackgroundEventHandler.java]: you don't really need this - option one: have kafkaConsumerMetrics and shareConsumerMetrics to impelment a common interface and do if(metrics instanceOf [one of the metrics class]) 

option two: either do null check or use optional.  Optional is better imo.

**Comment by AndrewJSchofield** [comment]
@brenden20 I think you're doing the right thing here with the share consumer, but it will need a small update to the KIP because those metrics are not listed.

**Comment by lianetm** [comment]
Closing this inactive draft. Changes are included in https://github.com/apache/kafka/pull/17199

**Pull request closed without merging**